projects/bootloader/src/dispatcher.c

```diff
@@ -0,0 +1,61 @@
+#include "dispatcher.h"
+
+#include <stdint.h>
+#include <string.h>
+
+#include "bootloader_can.h"
+#include "bootloader_datagram_defs.h"
+#include "can_datagram.h"
+
+#define MAX_DEST_NODES_SIZE_BYTES 255
+#define MAX_DATA_SIZE_BYTES 2048
+
+// callbacks and context
+static DispatcherCallback s_callback_map[NUM_BOOTLOADER_DATAGRAMS] = { NULL };
+static void *s_context_map[NUM_BOOTLOADER_DATAGRAMS] = { NULL };
+
+// setup datagram rx config
+static void prv_dispatch(void);
+static uint8_t s_destination_nodes[MAX_DEST_NODES_SIZE_BYTES] = { 0 };
+static uint8_t s_data[MAX_DATA_SIZE_BYTES] = { 0 };
+static CanDatagramRxConfig datagram_rx = {
+  .destination_nodes = s_destination_nodes,
+  .data = s_data,
+  // .node_id = 0, // node id is set at dispatcher_init
+  .rx_cmpl_cb = prv_dispatch,
+};
+
+// when a datagram rx is finished, the corresponding registered callback for that datagram id
+// is called with the datagram data and registered context
+static void prv_dispatch(void) {
+  BootloaderDatagramId id = datagram_rx.dgram_type;
+  if (id < NUM_BOOTLOADER_DATAGRAMS && s_callback_map[id] != NULL) {
+    s_callback_map[id](s_data, datagram_rx.data_len, s_context_map[id]);
+  }
```

Comment:
Since this doesn't return a statuscode, might be good to add a LOG_ERROR letting us know that an invalid ID or function has been called

Reply:
what module is LOG_ERROR in? Is it "log.h" with LOG_WARN or LOG_DEBUG ?